### ISSUE TITLE
Incorrect value of BenchmarkDotNet.Toolchains.DotNetCli.NetCoreAppSettings.Default

### ISSUE THREAD

**Issue by AndreyAkinshin** [dotnet/BenchmarkDotNet#1162]
Let's say that we have a `<TargetFrameworks>net461;netcoreapp2.1</TargetFrameworks>` project and we want to execute .NET Core benchmark under the `net461` TFM. In this case, `BenchmarkDotNet.Toolchains.DotNetCli.NetCoreAppSettings.Default` which will be used for TFM in the generated benchmark. Currently, it's `NetCoreApp20` which leads to a compilation error in the generated benchmark (because the base project doesn't support `netcoreapp2.0`).

Because of this problem, some of the Windows tests turned red after removing the `netcoreapp2.0` TFM from `BenchmarkDotNet.IntegrationTests` (see https://github.com/dotnet/BenchmarkDotNet/issues/1141#issuecomment-495259857, 6f91ea59474bf6cce2d7ff8de422854cb4a61b06).

Relevant builds:

* https://dev.azure.com/dotnet/BenchmarkDotNet/_build/results?buildId=17597
* https://ci.appveyor.com/project/dotnetfoundation/benchmarkdotnet/builds/24762511

Red tests:

![image](https://user-images.githubusercontent.com/2259237/58324454-903d2600-7e2f-11e9-8c1c-ea59f81c7a3e.png)

@adamsitnik do you have any ideas what we can do with this problem?

**Comment by adamsitnik** [comment]
The default is used only when the user runs the host process as full .NET Framework and does not specify which .NET Core exactly to run. Example:

```cmd
dotnet run -c Release -f net461 --runtimes core clr
```

.NET Core 2.0 is EOL so maybe we should just set the default to .NET Core 2.1?

**Comment by AndreyAkinshin** [comment]
@adamsitnik you are suggesting a workaround for this particular case, but I want to handle the general case.
Let's say we have a `<TargetFrameworks>net461;netcoreapp2.2</TargetFrameworks>` project and we want to execute .NET Core benchmark under the `net461`: the problem will be the same with the new default.

> and does not specify which .NET Core exactly to run

Actually, it's the most typical use case when you specify all the job characteristics in C# (e.g., `ManualConfig.Create(DefaultConfig.Instance).With(Job.RyuJitX64).With(Job.Core)`)

**Comment by adamsitnik** [comment]
I don't think that there is a good solution to this problem. We could parse the project file, but what if the project targets multiple TFMs? 

Which should be the default?

```xml
<TargetFrameworks>net461;netcoreapp3.0;netcoreapp2.2</TargetFrameworks>
```

Or uses an MSBuild Property?

```xml
<TFMS>net461;netcoreapp3.0;netcoreapp2.2</TFMS>
<TargetFrameworks>$(TFMS)</TargetFrameworks>
```

Or uses an MSBuild Condition?

```xml
<TargetFrameworks Condition="'$(TargetFrameworks)' == '' AND '$(OS)' == 'Windows_NT'">net461;netcoreapp2.1;netcoreapp2.2;netcoreapp3.0</TargetFrameworks>
    <TargetFrameworks Condition="'$(TargetFrameworks)' == ''">netcoreapp2.1;netcoreapp2.2;netcoreapp3.0</TargetFrameworks>
```

**Comment by adamsitnik** [comment]
Maybe we should make .NET Core 2.1 the default and add error handling for it?

So when the user runs as Full Framework and uses Job.Core or --runtimes core and fails we print a nice explanation?

**Comment by AndreyAkinshin** [comment]
> We could parse the project file, but what if the project targets multiple TFMs?

Agree, we can't get the full list of TFMs without honest MSBuild project evaluation.

> and add error handling for it?

Error handling is always a good thing to add. Currently, it's pretty hard to figure out what's going on.

---

Here is another idea of how we can resolve the problem. We can execute `dotnet --list-runtimes` and get the latest locally available .NET Core version. If we are trying to execute .NET Core benchmark from .NET Framework host process, we can use the latest version. What do you think?

**Comment by adamsitnik** [comment]
> If we are trying to execute .NET Core benchmark from .NET Framework host process, we can use the latest version. What do you think?

It's definitely a better idea, because it's always going to work. But it has one side effect. Users who target 2.1 but have 3.0 installed might get "better" results when running on 3.0 which has more optimizations. This might lead to a confusion and/or wrong results?

Speaking of that, maybe we should just require the users to specify the version in an explicit way and make it 100% explicit about what is executed?

```
[CoreJob("netcoreapp3.0")]
Job.Core21
--runtimes netcoreapp2.1 net472
```

@AndreyAkinshin what do you think? imho the value that we provide is the possibility to run the benchmarks for multiple runtimes which saves the users a LOT of time, requiring them to specify TFM in explicit way should be OK and would save us a lot of trouble.

**Comment by adamsitnik** [comment]
@AndreyAkinshin what do you think about requiring the users to specify the TFM? it would help us to avoid issues like https://github.com/dotnet/BenchmarkDotNet/issues/1170#issuecomment-497116717

**Comment by AndreyAkinshin** [comment]
@adamsitnik 
There are two things that I don't want to do:
1. Force users to write *magic string* (like `"netcoreapp3.0"`) manually. We can support such an option, but the recommended API shouldn't require such strings.
2. Force users to *always* write TFM. The most simple BenchmarkDotNet use case (a user creates an applications, write benchmark, run it) should be possible without additional requirements.

I think we can introduce an enum like this:
```cs
enum Tfm
{
    LatestStable, LatestPreview, NetCoreApp21, NetCoreApp22, NetCoreApp30
}
```
Next, it can be used in `CoreJob` like this:
```
[CoreJob(Tfm.LatestStable)]
[CoreJob(Tfm.NetCoreApp30)]
```
We can use `Tfm.LatestStable` as the default value.

**Comment by adamsitnik** [comment]
> Force users to write magic string

@AndreyAkinshin you are right, enum is much better than a magic string

> Force users to always write TFM

In my proposal, we would force the users to always write the TFM only if they want to compare multiple runtimes. So the change would affect only the users who compare .NET vs Core vs Mono vs CoreRT etc. 

I have taken a moment and looked for the issues that hit this specific problem:

- https://github.com/dotnet/BenchmarkDotNet/issues/1056
- https://github.com/dotnet/BenchmarkDotNet/issues/993
- https://github.com/dotnet/BenchmarkDotNet/issues/1182
- https://github.com/dotnet/BenchmarkDotNet/issues/1115

**Comment by adamsitnik** [comment]
I don't remember which PR exactly did fix it, but it was part of 0.12.0. Closing
